SignalServiceKit/src/Messages/OWSMessageUtils.m

```diff
@@ -94,6 +102,138 @@ - (NSUInteger)unreadMessagesInThread:(TSThread *)thread
+                             transaction:transaction];
+}
+
++ (nullable TSQuotedMessage *)quotedMessageForMessage:(TSMessage *)message
+                                             authorId:(NSString *)authorId
+                                               thread:(TSThread *)thread
+                                          transaction:(YapDatabaseReadTransaction *)transaction
+{
+    OWSAssert(message);
+    OWSAssert(authorId.length > 0);
+    OWSAssert(thread);
+    OWSAssert(transaction);
+
+    uint64_t timestamp = message.timestamp;
+    NSString *_Nullable body = message.body;
+    BOOL hasText = body.length > 0;
+    BOOL hasAttachment = NO;
+    NSString *_Nullable sourceFilename = nil;
+    NSData *_Nullable thumbnailData = nil;
+    NSString *_Nullable contentType = nil;
+
+    if (message.attachmentIds.count > 0) {
+        NSString *attachmentId = message.attachmentIds[0];
+        TSAttachment *_Nullable attachment =
+            [TSAttachment fetchObjectWithUniqueID:attachmentId transaction:transaction];
+        if (attachment) {
+            sourceFilename = attachment.sourceFilename;
+            contentType = attachment.contentType;
+            // Try to generate a thumbnail, if possible.
+            thumbnailData = [self thumbnailDataForAttachment:attachment];
+            hasAttachment = YES;
+        }
+    }
+
+    if (!hasText && !hasAttachment) {
+        OWSFail(@"%@ quoted message has neither text nor attachment", self.logTag);
+        return nil;
+    }
+
+    TSQuotedMessage *quotedMessage = [[TSQuotedMessage alloc] initWithTimestamp:timestamp
+                                                                       authorId:authorId
+                                                                           body:body
+                                                                 sourceFilename:sourceFilename
+                                                                  thumbnailData:thumbnailData
+                                                                    contentType:contentType];
+    return quotedMessage;
+}
+
++ (nullable NSData *)thumbnailDataForAttachment:(TSAttachment *)attachment
+{
+    OWSAssert(attachment);
+
+    // Try to generate a thumbnail, if possible.
+    if (![attachment isKindOfClass:[TSAttachmentStream class]]) {
+        return nil;
+    }
+
+    TSAttachmentStream *attachmentStream = (TSAttachmentStream *)attachment;
+    UIImage *_Nullable attachmentImage = [attachmentStream image];
+    if (!attachmentImage) {
+        return nil;
+    }
+
+    CGSize attachmentImageSizePx;
+    attachmentImageSizePx.width = CGImageGetWidth(attachmentImage.CGImage);
+    attachmentImageSizePx.height = CGImageGetHeight(attachmentImage.CGImage);
+    if (attachmentImageSizePx.width <= 0 || attachmentImageSizePx.height <= 0) {
+        DDLogError(@"%@ attachment thumbnail has invalid size.", self.logTag);
+        return nil;
+    }
+
+    const int kMaxThumbnailSizePx = 100;
```

Comment:
We'll have to reconcile this with where design lands.

Reply:
Yup, added a card.